### ISSUE TITLE
Windows to Android x64 cross-compilation

### ISSUE THREAD

**Issue by 999pingGG** [LuaJIT/LuaJIT#770]
I'm trying to compile for x86_64 Android using this command under Windows (shortened for [some] brevity):
```
mingw32-make
  HOST_CC="clang.exe -m64"
  CROSS="<ndk>\22.1.7171670\toolchains\llvm\prebuilt\windows-x86_64\bin\x86_64-linux-android-"
  TARGET_FLAGS="--sysroot <ndk>\22.1.7171670\toolchains\llvm\prebuilt\windows-x86_64\sysroot -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -O2 -DNDEBUG -Wl,-soname,libluajit.so -Wno-unused-command-line-argument -march=x86-64"
  TARGET_CC="<ndk>\22.1.7171670\toolchains\llvm\prebuilt\windows-x86_64\bin\x86_64-linux-android21-clang"
  STATIC_CC="<ndk>\22.1.7171670\toolchains\llvm\prebuilt\windows-x86_64\bin\x86_64-linux-android21-clang"
  DYNAMIC_CC="<ndk>\22.1.7171670\toolchains\llvm\prebuilt\windows-x86_64\bin\x86_64-linux-android21-clang -fPIC"
  TARGET_LD=<ndk>\22.1.7171670\toolchains\llvm\prebuilt\windows-x86_64\bin\x86_64-linux-android21-clang
  TARGET_SYS=Linux
  TARGET_AR="<ndk>\22.1.7171670\toolchains\llvm\prebuilt\windows-x86_64\bin\x86_64-linux-android-ar rcus"`
```

But I get the error in the title. At file `lj_err.c`, line 469 an `#error` triggers because both `#if !LJ_TARGET_ARM` and `#if LJ_TARGET_X64` are true. Was it assumed that any non-arm platform is not Android? I just comment out the `#error` and the build succeeds. The `#error` was added in commit e131936. Will anything go wrong if I use the resulting `libluajit.so`? Anyway, the compile-time check should be fixed but I'm not sure about how to do that properly.

**Comment by MikePall** [comment]
No, you're getting the error because `LUAJIT_UNWIND_EXTERNAL` is not defined. The check in `src/Makefile`, line 339 probably does not work on a Windows host.

In fact, cross-compilation from a Windows host using the supplied Makefiles is not a supported (or tested) configuration.

You could add that flag manually to `TARGET_FLAGS`. But better verify the resulting library has proper error handling.

**Comment by 999pingGG** [comment]
Thank you for your answer, I've now compiled for all Android archs using a Debian vm and it is much better now, I don't even get any compiler warnings unlike Windows! I will stop using the hacky ways that I came up with in the past because at the begginning I had trouble trying to get it to work.

There is another bug though: The Makefile didn't work at first. After some research turns out the NDK version r23b broke the build system yet again. They changed `ar` and `strip` locations, those are still at `toolchains\llvm\prebuilt\linux-x86_64\bin\`, but they no longer have the `$(CROSS)` prefix, now there is a single `llvm-ar` and `llvm-strip` for all archs.
